### PR TITLE
feat: add support for exception thrown in async method

### LINKED ISSUE #42: onThrowLogger doesn't handle async exceptions
First of all, thanks for addressing the cross-cutting concerns problem in JS, much appreciate the amazing work.

Here's my code:

```js
const { beforeMethod, afterMethod, onThrowOfMethod, Wove } = require('aspect.js');

class LoggerAspect {
    @beforeMethod({
        methodNamePattern: /.*/,
        classNamePattern: /.*/
    })
    beforeLogger(meta, ...args) {
        console.log(`Starting method ${meta.method.name} with arguments: ${args.join(', ')}`);
    }
    @afterMethod({
        methodNamePattern: /.*/,
        classNamePattern: /.*/
    })
    afterLogger(meta, ...args) {
        console.log(`Successfully completed method ${meta.method.name} with arguments: ${args.join(', ')}`);
    }
    @onThrowOfMethod({
        methodNamePattern: /.*/,
        classNamePattern: /.*/
    })
    onThrowLogger(meta, ...args) {
        console.log(`Error during the invocation of method ${meta.method.name} with arguments: ${args.join(', ')}`);
        console.log('Exception: ', meta.method.exception);
    }
}

@Wove({ bar: 42 })
class MyClass {
    myMethodX(x) {
        console.log(`myMethodX: ${x}.`);
        throw "Catching from myMethodX successfully";
    }
    async myMethodY(y) {
        console.log(`myMethodY: ${y}.`);
        throw "This exception from async myMethodY is not caught and not logged!";
        //return await Promise.reject("This will throw an exception that won't be caught or logged either!");
        //return Promise.reject("This rejection won't be caught or logged either!");
    }
}

const o = new MyClass()

console.log("Will invoke myMethodX with xxx");
o.myMethodX("xxx")
console.log("Will invoke myMethodY with yyy");
o.myMethodY("yyy")
```

Here's the output:
```bash
Will invoke myMethodX with xxx
Starting method myMethodX with arguments: xxx
myMethodX: xxx.
Error during the invocation of method myMethodX with arguments: xxx
Exception:  Catching from myMethodX successfully
Will invoke myMethodY with yyy
Starting method myMethodY with arguments: yyy
myMethodY: yyy.
Successfully completed method myMethodY with arguments: yyy

/usr/local/lib/node_modules/serverless/lib/classes/Error.js:93
    throw new Error(e);
    ^

Error: This exception from async myMethodY is not caught and not logged!
    at module.exports.logError (/usr/local/lib/node_modules/serverless/lib/classes/Error.js:93:11)
    at process.on (/usr/local/lib/node_modules/serverless/bin/serverless:12:41)
    at emitTwo (events.js:106:13)
    at process.emit (events.js:194:7)
    at emitPendingUnhandledRejections (internal/process/promises.js:85:22)
    at runMicrotasksCallback (internal/process/next_tick.js:67:9)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickDomainCallback (internal/process/next_tick.js:128:9)
```
The exception in the async method is not caught, instead the following log line is produced:
`Successfully completed method myMethodY with arguments: yyy`

Is this a bug or a feature?

### PR DESCRIPTION
**Pull request by mgechev**
Fix #42

### CODE REVIEW

File: lib/src/advices/async_advices.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import {AsyncAdvice} from '../core/advice';
+import {Metadata} from '../core/metadata';
+
+
```
Comment (SamVerschueren) [inline lib/src/advices/async_advices.ts]: Remove line :)

File: test/advices/async_advices.spec.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
+import {Metadata, MethodMetadata, Wove, resetRegistry} from '../../lib/src/core';
+import {Advice} from '../../lib/src/core/advice';
+import * as SyncAdvices from '../../lib/src/advices';
+
+import { afterMethod, asyncOnThrowOfMethod, onThrowOfMethod, aroundMethod } from '../../lib/index';
```
Comment (SamVerschueren) [inline test/advices/async_advices.spec.ts]: Remove spaces between brackets and names `{afterMethod` to be consistent.

**Comment by mgechev** [comment]
@SamVerschueren thanks for taking a look. Fixed.

File: lib/src/core/advice.ts [hunk 3]
```diff
@@ -9,11 +9,29 @@ export abstract class Advice {
...
+}
+
+export abstract class AsyncAdvice {
+  constructor(public context: Object, public advice: Function) {}
+
+  public async abstract wove(target: Function, metadata: Metadata);
+
+  public async invoke(target: any, metadata: Metadata) {
+    if (target.__woven__) {
+      return await (metadata.method.result = target.bind(this.context, metadata).apply(null, metadata.method.args));
```
Comment (SamVerschueren) [inline lib/src/core/advice.ts]: I'm going over the code again. I'm wondering, why don't we just do this here

```
if (target.__woven__) {
    metadata.method.result = await target.bind(this.context, metadata).apply(null, metadata.method.args);
} else if (metadata.method.proceed) {
    metadata.method.result = await target.apply(metadata.method.context, metadata.method.args));
}

return metadata.method.result;
```

Might be a a good reason for that though. But if not, I think this is more readable than what we have now. What do you think?

**Comment by mgechev** [comment]
@SamVerschueren let me know if you find something else.

**Comment by SamVerschueren** [comment]
I checked everything again. It looks good to me! The only small remark I have is that it might be a good idea of we rewrote [this block](https://github.com/mgechev/aspect.js/blob/master/lib/src/core/advice.ts#L9-L17) the same way as what we have with the async advices in this PR.

Great work @mgechev!

**Pull request merged**